syntax/crystal.vim

```diff
@@ -100,7 +126,12 @@ syn match  crystalSymbol		"\%([{(,]\_s*\)\@<=\l\w*[!?]\=::\@!"he=e-1
 syn match  crystalSymbol		"[]})\"':]\@<!\%(\h\|[^\x00-\x7F]\)\%(\w\|[^\x00-\x7F]\)*[!?]\=:\s\@="he=e-1
 syn match  crystalSymbol		"\%([{(,]\_s*\)\@<=[[:space:],{]\l\w*[!?]\=::\@!"hs=s+1,he=e-1
 syn match  crystalSymbol		"[[:space:],{]\%(\h\|[^\x00-\x7F]\)\%(\w\|[^\x00-\x7F]\)*[!?]\=:\s\@="hs=s+1,he=e-1
-syn region crystalSymbol		start="[]})\"':]\@<!:\"" end="\"" skip="\\\\\|\\\"" contains=@crystalStringSpecial fold
+
+if s:foldable(':')
+  syn region crystalSymbol		start="[]})\"':]\@<!:\"" end="\"" skip="\\\\\|\\\"" contains=@crystalStringSpecial fold
+else
+  syn region crystalSymbol  start="[]})\"':]\@<!:\"" end="\"" skip="\\\\\|\\\"" contains=@crystalStringSpecial
+endif
```

Comment:
I think these `if` switches are not necessary. We can simply add `fold` attribute because it is not effective until user runs `:set foldmethod=syntax` (default value of `foldmethod` is `manual`).

Reply:
They're necessary for allowing the user to specify which groups that they want to be folded. For example, if `:` isn't in the user's `crystal_foldable_groups`, then we don't want `crystalSymbol` to be marked with `fold`, hence the check for `s:foldable(':')`.

Reply:
I see. I missed the point, thank you.

And one more point. Currently this patch introduces so much double works since it duplicates many syntax definitions.

I checked vim-ruby source and it defines its own `SynFold` command for avoiding the double works:

https://github.com/vim-ruby/vim-ruby/blob/master/syntax/ruby.vim#L65

Would you apply the same manner to this patch if we goes with dynamic folding groups?

Reply:
Yeah, you're right. I was using an outdated version of `vim-ruby` for reference when I made these changes. I'll see if I can copy the way `vim-ruby` does it now.

As an aside, I noticed that `vim-ruby`'s current syntax file has a lot of syntax groups that `vim-crystal` doesn't have, probably because `vim-ruby` has been getting a lot of updates lately. I think it would be a lot easier to port features from `vim-ruby` to `vim-crystal` in the feature if the source files were as similar as possible.

Would you be willing to review a separate PR that refactors the syntax file to be as close to `vim-ruby`'s as possible (including folding), or should I just stick to folding for now?

Reply:
Yeah, I think we should focus on folding in this PR and this PR should be merged into master at first. If you want the improvements added to vim-ruby recently in this plugin, it's more than welcome to have a separate PR for them.